File: tools/richos-service/companion-macos/Sources/RichOSCompanionCore/SessionContract.swift
import Foundation

/// Builds the frozen capture->pipeline contract (schemaVersion 2) that P1's pipeline consumes.
///
/// The single most load-bearing interface in the system (architecture §3 / P1 README "The frozen
/// capture->pipeline contract"). The macOS companion is "just another producer" of the SAME
/// session directory the extension writes — so the pipeline (§4) and reliability model (§6) are
/// surface-independent. Every field name/shape below is matched to what `lib/contract.js`,
/// `lib/pipeline.js`, `lib/reconcile.js` and `merge.js#renderMarkdown` read.
public enum SessionContract {

    /// The frozen `capture.source` enum value for this surface (P1 `CAPTURE_SOURCE.macos`).
    /// NB: the frozen enum is `desktop-companion-macos` — the task brief's shorthand "companion-macos"
    /// resolves to this exact literal, which is what `contract.js` and the pipeline expect.
    public static let captureSource = "desktop-companion-macos"
    public static let captureMethod = "coreaudio-tap+mic"

    /// Portable, `:`-free, tz-unambiguous timestamp — byte-identical to the extension's
    /// `session.js#stampFor`: `new Date(ms).toISOString().replace(/\.\d+Z$/, 'Z').replace(/:/g, '-')`.
    public static func stampFor(_ epochMs: Int64) -> String {
        let date = Date(timeIntervalSince1970: Double(epochMs) / 1000.0)
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let c = cal.dateComponents([.year, .month, .day, .hour, .minute, .second], from: date)
        return String(
            format: "%04d-%02d-%02dT%02d-%02d-%02dZ",
            c.year!, c.month!, c.day!, c.hour!, c.minute!, c.second!
        )
    }

    /// `${stamp}--${platformId}--${slug}` — matches `session.js#sessionDirName`.
    public static func sessionDirName(startedAt: Int64, platformId: String, slug: String) -> String {
        "\(stampFor(startedAt))--\(platformId)--\(slug)"
    }

    public static func audioPartFile(_ part: Int) -> String {
        String(format: "audio-part-%02d.wav", part)
    }

    /// A finished audio part's accounting, as it appears in `session.json.audio.parts[]`.
    public struct AudioPart {
        public let part: Int
        public let bytes: Int
        public let firstChunkAt: Int64?
        public let lastChunkAt: Int64?
        public init(part: Int, bytes: Int, firstChunkAt: Int64?, lastChunkAt: Int64?) {
            self.part = part
            self.bytes = bytes
            self.firstChunkAt = firstChunkAt
            self.lastChunkAt = lastChunkAt
        }
    }

    /// Rolled-up health accounting for `session.json.health` (mirrors `session.js#accrueHealth`).
    public struct HealthSummary {
        public var heartbeats = 0
        public var greenSeconds = 0
        public var amberSeconds = 0
        public var redSeconds = 0
        public var worstLevel = "green"
        public var recordsWritten = 0
        public init() {}
    }

    public struct Params {
        public var sessionId: String
        public var startedAt: Int64
        public var sampleRate: Int
        public var chunkMs: Int
        public var micEnabled: Bool
        public var captureTarget: String
        public var platformId: String
        public var platformLabel: String
        public var platformSlug: String
        public var companionVersion: String
        public var processHint: String?
        public init(
            sessionId: String, startedAt: Int64, sampleRate: Int, chunkMs: Int, micEnabled: Bool,
            captureTarget: String, platformId: String, platformLabel: String, platformSlug: String,
            companionVersion: String, processHint: String?
        ) {
            self.sessionId = sessionId
            self.startedAt = startedAt
            self.sampleRate = sampleRate
            self.chunkMs = chunkMs
            self.micEnabled = micEnabled
            self.captureTarget = captureTarget
            self.platformId = platformId
            self.platformLabel = platformLabel
            self.platformSlug = platformSlug
            self.companionVersion = companionVersion
            self.processHint = processHint
        }
    }

    /// The full v2 `session.json` object. `status`/`endedAt`/`audio`/`health` change between the
    /// call-START write (`open`, empty audio) and the call-CLOSE write (`closed`, populated) — the
    /// never-silent inversion (§6.1): the record is on disk from second zero, so a call that
    /// captured nothing is a loud `open`/empty anomaly, never an absence.
    public static func sessionJSON(
        _ p: Params,
        status: String,
        endedAt: Int64?,
        audioParts: [AudioPart],
        health: HealthSummary
    ) -> JSON {
        let bytesTotal = audioParts.reduce(0) { $0 + $1.bytes }

        let partsJSON: [JSON] = audioParts.map { part in
            .object([
                ("part", .int(part.part)),
                ("file", .string(audioPartFile(part.part))),
                ("bytes", .int(part.bytes)),
                ("firstChunkAt", part.firstChunkAt.map { JSON.int(Int($0)) } ?? .null),
                ("lastChunkAt", part.lastChunkAt.map { JSON.int(Int($0)) } ?? .null),
                ("written", .bool(true)),
            ])
        }

        return .object([
            ("schemaVersion", .int(2)),
            ("sessionId", .string(p.sessionId)),
            ("dir", .string(p.sessionId)),
            ("status", .string(status)),
            ("producer", .object([
                ("product", .string("RichOS macOS companion")),
                ("module", .string("call-capture")),
                ("companionVersion", .string(p.companionVersion)),
            ])),
            ("platform", .object([
                ("id", .string(p.platformId)),
                ("label", .string(p.platformLabel)),
                ("slug", .string(p.platformSlug)),
            ])),
            ("startedAt", .int(Int(p.startedAt))),
            ("endedAt", endedAt.map { JSON.int(Int($0)) } ?? .null),
            ("capture", .object([
                ("source", .string(captureSource)),
                ("method", .string(captureMethod)),
                ("channels", .object([
                    ("left", .string("microphone (me)")),
                    ("right", .string("system/tap (everyone else)")),
                ])),
                ("captureTarget", .string(p.captureTarget)),
                ("sampleRate", .int(p.sampleRate)),
                ("container", .string("audio/wav;codecs=pcm_s16le")),
                ("micEnabled", .bool(p.micEnabled)),
                ("chunkMs", .int(p.chunkMs)),
            ])),
            ("audio", .object([
                ("parts", .array(partsJSON)),
                ("bytesTotal", .int(bytesTotal)),
                ("chunkCount", .int(audioParts.count)),
            ])),
            ("health", .object([
                ("heartbeats", .int(health.heartbeats)),
                ("greenSeconds", .int(health.greenSeconds)),
                ("amberSeconds", .int(health.amberSeconds)),
                ("redSeconds", .int(health.redSeconds)),
                ("worstLevel", .string(health.worstLevel)),
                ("recordsWritten", .int(health.recordsWritten)),
            ])),
            ("alerts", .array([])),
            ("recovery", .array([])),
            // §3.4: a desktop-app companion has NO platform captions. Present-from-birth, empty.
            ("captions", .object([
                ("available", .bool(false)),
                ("adapter", .null),
                ("adapterVersion", .null),
                ("count", .int(0)),
                ("speakers", .array([])),
                ("degraded", .bool(false)),
            ])),
            ("mode", .string("full")),
            ("notes", .array([])),
            // §5.4 dedup handshake block — a lone companion session owns itself.
            ("ownership", .object([
                ("ownerSurface", .string(captureSource)),
                ("supersedes", .null),
                ("processHint", p.processHint.map { JSON.string($0) } ?? .null),
            ])),
            // Written by the pipeline; born `pending` so a never-run pipeline is visible on disk.
            ("pipeline", .object([
                ("status", .string("pending")),
                ("model", .null),
                ("modelRuns", .array([])),
                ("ffmpegVersion", .null),
                ("whisperVersion", .null),
                ("loroCorrection", .object([
                    ("applied", .bool(false)),
                    ("entitiesVersion", .null),
                    ("corrections", .int(0)),
                ])),
            ])),
        ])
    }

    /// One `health.ndjson` row (§3.3 absolute-`t` keyed; §6.2 second-by-second signals). Contract:
    /// "per-channel levels, byte growth, state." Named for the companion's two sources (mic / system
    /// tap) rather than the extension's (mic / tab).
    public struct HealthSample {
        public var sessionId: String
        public var t: Int64
        public var micRms: Double
        public var sysRms: Double
        public var micRmsMean: Double
        public var sysRmsMean: Double
        public var bytesTotal: Int
        public var bytesDelta: Int
        public var part: Int
        public var tapRunning: Bool
        public var micRunning: Bool
        public var level: String
        public var problems: [String]
        public init(
            sessionId: String, t: Int64, micRms: Double, sysRms: Double, micRmsMean: Double,
            sysRmsMean: Double, bytesTotal: Int, bytesDelta: Int, part: Int, tapRunning: Bool,
            micRunning: Bool, level: String, problems: [String]
        ) {
            self.sessionId = sessionId
            self.t = t
            self.micRms = micRms
            self.sysRms = sysRms
            self.micRmsMean = micRmsMean
            self.sysRmsMean = sysRmsMean
            self.bytesTotal = bytesTotal
            self.bytesDelta = bytesDelta
            self.part = part
            self.tapRunning = tapRunning
            self.micRunning = micRunning
            self.level = level
            self.problems = problems
        }

        public func json() -> JSON {
            .object([
                ("sessionId", .string(sessionId)),
                ("t", .int(Int(t))),
                ("micRms", .double(round6(micRms))),
                ("sysRms", .double(round6(sysRms))),
                ("micRmsMean", .double(round6(micRmsMean))),
                ("sysRmsMean", .double(round6(sysRmsMean))),
                ("bytesTotal", .int(bytesTotal)),
                ("bytesDelta", .int(bytesDelta)),
                ("part", .int(part)),
                ("tapRunning", .bool(tapRunning)),
                ("micRunning", .bool(micRunning)),
                ("level", .string(level)),
                ("problems", .array(problems.map { JSON.string($0) })),
            ])
        }

        private func round6(_ x: Double) -> Double {
            (x * 1_000_000).rounded() / 1_000_000
        }
    }
}

File: tools/richos-service/companion-windows/src/RichOSCompanionCore/SessionContract.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace RichOSCompanionCore
{
    /// <summary>
    /// Builds the frozen capture-&gt;pipeline contract (schemaVersion 2) that P1's pipeline consumes.
    ///
    /// The single most load-bearing interface in the system (architecture §3 / P1 README "The frozen
    /// capture-&gt;pipeline contract"). The Windows companion is "just another producer" of the SAME
    /// session directory the extension and the macOS companion write — so the pipeline (§4) and the
    /// reliability model (§6) are surface-independent. Every field name/shape below is matched to what
    /// <c>lib/contract.js</c>, <c>lib/pipeline.js</c>, <c>lib/reconcile.js</c> and
    /// <c>merge.js#renderMarkdown</c> read, and to the macOS companion's <c>SessionContract.swift</c>.
    /// </summary>
    public static class SessionContract
    {
        /// <summary>
        /// The frozen <c>capture.source</c> enum value for this surface (P1
        /// <c>CAPTURE_SOURCE.windows</c> in <c>lib/contract.js</c> = <c>desktop-companion-windows</c>).
        /// </summary>
        public const string CaptureSource = "desktop-companion-windows";

        /// <summary>Default method for all-system loopback (architecture §3.2 enum <c>wasapi-loopback+mic</c>).</summary>
        public const string MethodSystemLoopback = "wasapi-loopback+mic";

        /// <summary>Method when scoped to a target process tree (architecture §5.3 process loopback).</summary>
        public const string MethodProcessLoopback = "wasapi-process-loopback+mic";

        /// <summary>
        /// Portable, <c>:</c>-free, tz-unambiguous timestamp — byte-identical to the extension's
        /// <c>session.js#stampFor</c> and the macOS companion's <c>stampFor</c>:
        /// <c>new Date(ms).toISOString().replace(/\.\d+Z$/, 'Z').replace(/:/g, '-')</c>.
        /// </summary>
        public static string StampFor(long epochMs)
        {
            var dt = DateTimeOffset.FromUnixTimeMilliseconds(epochMs).UtcDateTime;
            return string.Format(
                CultureInfo.InvariantCulture,
                "{0:D4}-{1:D2}-{2:D2}T{3:D2}-{4:D2}-{5:D2}Z",
                dt.Year, dt.Month, dt.Day, dt.Hour, dt.Minute, dt.Second);
        }

        /// <summary><c>${stamp}--${platformId}--${slug}</c> — matches <c>session.js#sessionDirName</c>.</summary>
        public static string SessionDirName(long startedAt, string platformId, string slug)
            => $"{StampFor(startedAt)}--{platformId}--{slug}";

        public static string AudioPartFile(int part)
            => string.Format(CultureInfo.InvariantCulture, "audio-part-{0:D2}.wav", part);

        /// <summary>A finished audio part's accounting, as in <c>session.json.audio.parts[]</c>.</summary>
        public readonly struct AudioPart
        {
            public readonly int Part;
            public readonly int Bytes;
            public readonly long? FirstChunkAt;
            public readonly long? LastChunkAt;
            public AudioPart(int part, int bytes, long? firstChunkAt, long? lastChunkAt)
            {
                Part = part; Bytes = bytes; FirstChunkAt = firstChunkAt; LastChunkAt = lastChunkAt;
            }
        }

        /// <summary>Rolled-up health accounting for <c>session.json.health</c> (mirrors <c>session.js#accrueHealth</c>).</summary>
        public sealed class HealthSummary
        {
            public int Heartbeats;
            public int GreenSeconds;
            public int AmberSeconds;
            public int RedSeconds;
            public string WorstLevel = "green";
            public int RecordsWritten;
        }

        public sealed class Params
        {
            public string SessionId;
            public long StartedAt;
            public int SampleRate;
            public int ChunkMs;
            public bool MicEnabled;
            public string CaptureTarget;
            public string Method;
            public string PlatformId;
            public string PlatformLabel;
            public string PlatformSlug;
            public string CompanionVersion;
            public string? ProcessHint;
            /// <summary>Set when this session PROMOTES over a dead browser call (§5.4 failover).</summary>
            public string? Supersedes;

            public Params(
                string sessionId, long startedAt, int sampleRate, int chunkMs, bool micEnabled,
                string captureTarget, string method, string platformId, string platformLabel,
                string platformSlug, string companionVersion, string? processHint, string? supersedes = null)
            {
                SessionId = sessionId;
                StartedAt = startedAt;
                SampleRate = sampleRate;
                ChunkMs = chunkMs;
                MicEnabled = micEnabled;
                CaptureTarget = captureTarget;
                Method = method;
                PlatformId = platformId;
                PlatformLabel = platformLabel;
                PlatformSlug = platformSlug;
                CompanionVersion = companionVersion;
                ProcessHint = processHint;
                Supersedes = supersedes;
            }
        }

        /// <summary>
        /// The full v2 <c>session.json</c> object. <c>status</c>/<c>endedAt</c>/<c>audio</c>/<c>health</c>
        /// change between the call-START write (<c>open</c>, empty audio) and the call-CLOSE write
        /// (<c>closed</c>, populated) — the never-silent inversion (§6.1): the record is on disk from
        /// second zero, so a call that captured nothing is a loud <c>open</c>/empty anomaly, never an
        /// absence.
        /// </summary>
        public static Json SessionJson(
            Params p, string status, long? endedAt, IReadOnlyList<AudioPart> audioParts, HealthSummary health)
        {
            int bytesTotal = 0;
            var partsJson = new List<Json>(audioParts.Count);
            foreach (var part in audioParts)
            {
                bytesTotal += part.Bytes;
                partsJson.Add(Json.Obj(
                    ("part", Json.Int(part.Part)),
                    ("file", Json.Str(AudioPartFile(part.Part))),
                    ("bytes", Json.Int(part.Bytes)),
                    ("firstChunkAt", part.FirstChunkAt.HasValue ? Json.Int(part.FirstChunkAt.Value) : Json.Null),
                    ("lastChunkAt", part.LastChunkAt.HasValue ? Json.Int(part.LastChunkAt.Value) : Json.Null),
                    ("written", Json.Bool(true))));
            }

            return Json.Obj(
                ("schemaVersion", Json.Int(2)),
                ("sessionId", Json.Str(p.SessionId)),
                ("dir", Json.Str(p.SessionId)),
                ("status", Json.Str(status)),
                ("producer", Json.Obj(
                    ("product", Json.Str("RichOS Windows companion")),
                    ("module", Json.Str("call-capture")),
                    ("companionVersion", Json.Str(p.CompanionVersion)))),
                ("platform", Json.Obj(
                    ("id", Json.Str(p.PlatformId)),
                    ("label", Json.Str(p.PlatformLabel)),
                    ("slug", Json.Str(p.PlatformSlug)))),
                ("startedAt", Json.Int(p.StartedAt)),
                ("endedAt", endedAt.HasValue ? Json.Int(endedAt.Value) : Json.Null),
                ("capture", Json.Obj(
                    ("source", Json.Str(CaptureSource)),
                    ("method", Json.Str(p.Method)),
                    ("channels", Json.Obj(
                        ("left", Json.Str("microphone (me)")),
                        // Descriptive only — the pipeline maps by channel POSITION (L->me, R->others via
                        // ffmpeg channelsplit), never by this string. Honest about the Windows mechanism.
                        ("right", Json.Str("system/loopback (everyone else)")))),
                    ("captureTarget", Json.Str(p.CaptureTarget)),
                    ("sampleRate", Json.Int(p.SampleRate)),
                    ("container", Json.Str("audio/wav;codecs=pcm_s16le")),
                    ("micEnabled", Json.Bool(p.MicEnabled)),
                    ("chunkMs", Json.Int(p.ChunkMs)))),
                ("audio", Json.Obj(
                    ("parts", Json.Arr(partsJson)),
                    ("bytesTotal", Json.Int(bytesTotal)),
                    ("chunkCount", Json.Int(audioParts.Count)))),
                ("health", Json.Obj(
                    ("heartbeats", Json.Int(health.Heartbeats)),
                    ("greenSeconds", Json.Int(health.GreenSeconds)),
                    ("amberSeconds", Json.Int(health.AmberSeconds)),
                    ("redSeconds", Json.Int(health.RedSeconds)),
                    ("worstLevel", Json.Str(health.WorstLevel)),
                    ("recordsWritten", Json.Int(health.RecordsWritten)))),
                ("alerts", Json.Arr()),
                ("recovery", Json.Arr()),
                // §3.4: a desktop-app companion has NO platform captions. Present-from-birth, empty.
                ("captions", Json.Obj(
                    ("available", Json.Bool(false)),
                    ("adapter", Json.Null),
                    ("adapterVersion", Json.Null),
                    ("count", Json.Int(0)),
                    ("speakers", Json.Arr()),
                    ("degraded", Json.Bool(false)))),
                ("mode", Json.Str("full")),
                ("notes", Json.Arr()),
                // §5.4 dedup handshake block — a lone companion session owns itself; `supersedes` is
                // set when this session PROMOTES over a dead browser-owned call (failover).
                ("ownership", Json.Obj(
                    ("ownerSurface", Json.Str(CaptureSource)),
                    ("supersedes", Json.StrOrNull(p.Supersedes)),
                    ("processHint", Json.StrOrNull(p.ProcessHint)))),
                // Written by the pipeline; born `pending` so a never-run pipeline is visible on disk.
                ("pipeline", Json.Obj(
                    ("status", Json.Str("pending")),
                    ("model", Json.Null),
                    ("modelRuns", Json.Arr()),
                    ("ffmpegVersion", Json.Null),
                    ("whisperVersion", Json.Null),
                    ("loroCorrection", Json.Obj(
                        ("applied", Json.Bool(false)),
                        ("entitiesVersion", Json.Null),
                        ("corrections", Json.Int(0)))))));
        }

        /// <summary>
        /// One <c>health.ndjson</c> row (§3.3 absolute-<c>t</c> keyed; §6.2 second-by-second signals).
        /// Field names are IDENTICAL to the macOS companion's <c>HealthSample</c> so <c>health.ndjson</c>
        /// is a uniform cross-surface schema; on Windows <c>tapRunning</c> means "the WASAPI loopback
        /// capture client is delivering" and <c>sysRms</c> is the loopback (RIGHT) channel level.
        /// </summary>
        public sealed class HealthSample
        {
            public string SessionId = "";
            public long T;
            public double MicRms;
            public double SysRms;
            public double MicRmsMean;
            public double SysRmsMean;
            public int BytesTotal;
            public int BytesDelta;
            public int Part;
            public bool TapRunning;
            public bool MicRunning;
            public string Level = "green";
            public IReadOnlyList<string> Problems = Array.Empty<string>();

            public Json ToJson()
            {
                var probs = new List<Json>(Problems.Count);
                foreach (var s in Problems) probs.Add(Json.Str(s));
                return Json.Obj(
                    ("sessionId", Json.Str(SessionId)),
                    ("t", Json.Int(T)),
                    ("micRms", Json.Dbl(Round6(MicRms))),
                    ("sysRms", Json.Dbl(Round6(SysRms))),
                    ("micRmsMean", Json.Dbl(Round6(MicRmsMean))),
                    ("sysRmsMean", Json.Dbl(Round6(SysRmsMean))),
                    ("bytesTotal", Json.Int(BytesTotal)),
                    ("bytesDelta", Json.Int(BytesDelta)),
                    ("part", Json.Int(Part)),
                    ("tapRunning", Json.Bool(TapRunning)),
                    ("micRunning", Json.Bool(MicRunning)),
                    ("level", Json.Str(Level)),
                    ("problems", Json.Arr(probs)));
            }

            private static double Round6(double x) => Math.Round(x * 1_000_000.0) / 1_000_000.0;
        }
    }
}
